### PR TITLE
Task #200127: AWS Rekogonizition API POC to detect face similarities

### PR DESCRIPTION
**Pull request by rushi-tekdi**
-debug and solved error of associateFaceToUser: InvalidParameterException: in @aws-sdk/client-rekognition callback function
-debug and fixed user creation and AssociateFacesCommand 
-created a graphql query to fetch users whose accounts were already created under the AWS collection 
-pass this response of graphql search to AssociateFacesCommand 
-tested with first 5 users face index creation in aws face rekognition index and user data update in users table
-debug and fixed user face matching index similarity and SearchUsersByImageCommand errors 
-created a graphql query to fetch users whose attendance were already created and photo uploaded 
-pass this response of graphql search to SearchUsersByImageCommand 
-tested with ABC user attendance in aws face rekognition index and user data update in attendance table
-attendance id used in functional test is XYZ
-divided entire workflow in created 3 cron job .i.e., user creation in collection, face indexing of created users, mark attendance of face indexed and attendance photo uploaded users
-compare code shared by other devs and making changes according in new updated code
-changed env files and its dependent variables in code
-removed all success console log, added error.stack and proper return response while exception occurred
-solved required code smells in sonarcloud
-add custom cron job timer 
-first cron runs for each hour's 5th minute eg: 10:05am, 11::05am
-2nd cron runs for each hour's 15th minute eg: 10:15am, 11::15am
-3rd cron runs for each hour's 25th minute eg: 10:25am, 11::25am

# I have ensured that following `Pull Request Checklist` is taken care of before sending this PR

* [ ] Code is formatted as per format decided
* [ ] Updated acceptance criteria in tracker
* [ ] Updated test cases in test-cases-tracker
* [ ] Updated new API endpoint if any in common postman collection
* [ ] Updated DB changes in db-tracker if any

### CODE REVIEW

File: src/package.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
     "test:e2e": "jest --config ./test/jest-e2e.json"
   },
   "dependencies": {
+		"@aws-sdk/client-rekognition": "^3.410.0",
```
Comment (manojLondhe) [inline src/package.json:23]: use ^3.423.0 and rechk again

File: src/package.json [hunk 2]
```diff
@@ -28,8 +29,11 @@
     "@nestjs/core": "^9.0.0",
     "@nestjs/mapped-types": "^1.2.2",
     "@nestjs/platform-express": "^9.0.0",
+		"@nestjs/schedule": "^3.0.1",
     "@sentry/node": "^7.52.1",
     "atob": "^2.1.2",
+		"aws-sdk": "^2.1454.0",
+		"aws-sdk-js-codemod": "^0.17.7",
```
Comment (manojLondhe) [inline src/package.json:36]: this codemod may not be needed, remove and rechk once

File: src/src/services/aws-rekognition/aws-rekognition.service.ts [hunk 3]
```diff
@@ -0,0 +1,370 @@
+// import * as AWS from 'aws-sdk';
```
Comment (manojLondhe) [inline src/src/services/aws-rekognition/aws-rekognition.service.ts:1]: remove this commented code

File: src/src/services/aws-rekognition/aws-rekognition.service.ts [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+			return users;
+		} catch (error) {
+			console.log('getAllUsersOfCollection:', error, error.stack);
+			return [];
+		}
+	}
+
+	async createUsersInCollection(collectionId: string, userIds: any) {
+		try {
+			//only create first 10 users in aws collection
```
Comment (manojLondhe) [inline src/src/services/aws-rekognition/aws-rekognition.service.ts:106]: is comment valid? where is 10 coming from?

File: src/src/services/aws-rekognition/aws-rekognition.service.ts [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+			if (disassociateFaceResponse.DisassociatedFaces.length === 1)
+				response.success = true;
+			return response;
+		} catch (error) {
+			console.log('disassociatePhotoFromUser:', error, error.stack);
+			return response;
+		}
+	}
+
+	async deletePhotoFromCollection(collectionId: string, faceId: string) {
```
Comment (manojLondhe) [inline src/src/services/aws-rekognition/aws-rekognition.service.ts:211]: rename deletePhotoFromCollection =>  deleteFaceFromCollection

File: src/src/services/aws-rekognition/aws-rekognition.service.ts [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+	async deleteCollection(
+		collectionId: string,
+		userId: string,
+		faceId: string,
+	) {
+		try {
+			const deleteFaceParams = {
+				CollectionId: collectionId,
+				userId: this.prefixed + userId,
```
Comment (manojLondhe) [inline src/src/services/aws-rekognition/aws-rekognition.service.ts:355]: 1. rename `const deleteFaceParams to const deleteCollectionParams`

2. are these needed for deleting collection?

```
				userId: this.prefixed + userId,
				FaceIds: [faceId],
```

File: src/src/cron/cron.module.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { Module } from '@nestjs/common';
+import { UserCreationService } from './usercreation.service';
+import { FaceIndexingService } from './faceIndexing.service';
+import { MarkAttendanceService } from './markAttendance.service';
+import { AwsRekognitionModule } from '../services/aws-rekognition/aws-rekognition.module';
+import { HasuraModule } from '../services/hasura/hasura.module';
+@Module({
+	imports: [AwsRekognitionModule, HasuraModule],
+	providers: [
+		UserCreationService,
```
Comment (manojLondhe) [inline src/src/cron/cron.module.ts:10]: lets rename these services as

usercreation.service.ts => faUserIndexing.cron.ts
faceIndexing.service.ts => faFaceIndexing.cron.ts
markAttendance.service.ts => faAttendanceProcessing.cron.ts

UserCreationService => FaUserIndexingCron
FaceIndexingService => FaFaceIndexingCron
MarkAttendanceService => FaAttendanceProcessingCron

File: src/src/cron/usercreation.service.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+			);
+		} catch (error) {
+			console.log(
+				'Error occurred in createCollectionUsers.',
+				error,
+				error.stack,
+			);
+		}
+	}
+	async fetchAllUsersExceptCreated(limit) {
```
Comment (manojLondhe) [inline src/src/cron/usercreation.service.ts:65]: add empty line before starting new function

File: src/src/cron/usercreation.service.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+		}
+	}
+	async deleteCollection(
+		collectionId: string,
+		userId: string,
+		faceId: string,
+	) {
+		// Disassociate image from user
+		const photoDisassociated =
+			await this.awsRekognitionService.deleteCollection(
```
Comment (manojLondhe) [inline src/src/cron/usercreation.service.ts:98]: is it delete collection of diassociate face?

File: src/src/cron/usercreation.service.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+				collectionId,
+				userId,
+				faceId,
+			);
+		//console.log('photoDisassociated111------>>>>>', photoDisassociated);
+		let response = { success: false };
+		// Delete face from collection
+		if (photoDisassociated) {
+			const photoDeleted =
+				await this.awsRekognitionService.deleteCollection(
```
Comment (manojLondhe) [inline src/src/cron/usercreation.service.ts:108]: is it delete collection of delete face?

File: src/src/cron/faceIndexing.service.ts [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+		);*/
+
+		// We need to skip processing records wch were processed in past X hours
+		let dt = new Date();
+		let filterTimestamp = new Date(
+			dt.setHours(dt.getHours() - 4),
+		).toISOString();
+
+		const query = `
+				query MyQuery {
```
Comment (manojLondhe) [inline src/src/cron/faceIndexing.service.ts:207]: remove one level indent

File: src/src/cron/faceIndexing.service.ts [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+		userId: number,
+		{ fa_photos_last_processed_at },
+	) {
+		/*console.log(
+			'START - Mark fa last processed timestamp for user with: userId, as',
+			userId,
+		);*/
+
+		let updateQuery = `
+            mutation MyMutation {
```
Comment (manojLondhe) [inline src/src/cron/faceIndexing.service.ts:349]: increase indent by one level

File: src/src/cron/markAttendance.service.ts [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+			);
+		} catch (error) {
+			console.log('markAttendance:', error, error.stack);
+			return [];
+		}
+	}
+
+	async markProcessed(attendaceId: number) {
+		let updateQuery = `
+				mutation MyMutation {
```
Comment (manojLondhe) [inline src/src/cron/markAttendance.service.ts:156]: remove one level indent

File: src/src/cron/markAttendance.service.ts [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+				(await this.hasuraService.getData({ query: updateQuery })).data
+					.update_attendance_by_pk.id === attendaceId
+			);
+		} catch (error) {
+			console.log('markAttendance:', error, error.stack);
+			return [];
+		}
+	}
+
+	async markProcessed(attendaceId: number) {
```
Comment (manojLondhe) [inline src/src/cron/markAttendance.service.ts:154]: attendaceId => attendanceId

**Review by manojLondhe — CHANGES_REQUESTED** [review]

File: src/src/services/aws-rekognition/aws-rekognition.service.ts (hunk 6 above)
Comment (rushi-tekdi) [reply src/src/services/aws-rekognition/aws-rekognition.service.ts:355]: userId: this.prefixed + userId,	FaceIds: [faceId], not required in DeleteCollectionCommand
I read it on below documentation link
https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-rekognition/classes/deletecollectioncommand.html

Comment (rushi-tekdi) [reply src/src/services/aws-rekognition/aws-rekognition.service.ts:355]: I renamed const deleteFaceParams to const deleteCollectionParams. Also Removed userId, FaceId from deleteCollection method

File: src/src/services/aws-rekognition/aws-rekognition.service.ts (hunk 5 above)
Comment (rushi-tekdi) [reply src/src/services/aws-rekognition/aws-rekognition.service.ts:211]: done

File: src/package.json (hunk 2 above)
Comment (rushi-tekdi) [reply src/package.json:36]: I removed it.

File: src/package.json (hunk 1 above)
Comment (rushi-tekdi) [reply src/package.json:23]: I changed version.

File: src/src/cron/usercreation.service.ts (hunk 8 above)
Comment (rushi-tekdi) [reply src/src/cron/usercreation.service.ts:65]: i added empty line in all function starting

File: src/src/cron/cron.module.ts (hunk 7 above)
Comment (rushi-tekdi) [reply src/src/cron/cron.module.ts:10]: I renamed all as you suggested.

File: src/src/cron/usercreation.service.ts (hunk 9 above)
Comment (rushi-tekdi) [reply src/src/cron/usercreation.service.ts:98]: this code is created by parth gandhi. so didn't checked it. can I remove it?

File: src/src/cron/faceIndexing.service.ts (hunk 11 above)
Comment (rushi-tekdi) [reply src/src/cron/faceIndexing.service.ts:207]: done

File: src/src/cron/faceIndexing.service.ts (hunk 12 above)
Comment (rushi-tekdi) [reply src/src/cron/faceIndexing.service.ts:349]: done

File: src/src/cron/markAttendance.service.ts (hunk 14 above)
Comment (rushi-tekdi) [reply src/src/cron/markAttendance.service.ts:154]: done

**Pull request merged**